src/components/core/geoChart/GeoChartLegendRenderer.tsx

```diff
@@ -4,66 +4,190 @@ import Measure, { ContentRect, MeasuredComponentProps } from "react-measure";
+        return null;
+    }
+
+    return (
+        <IntlWrapper locale={locale}>
+            <IntlTranslationsProvider>
+                {(translationsProps: ITranslationsComponentProps) => {
+                    const { numericSymbols } = translationsProps;
+
+                    if (hasCategoryLegend) {
+                        return renderCategoryAndSizeLegend(props, numericSymbols, availableLegends);
+                    }
+
+                    if (hasColorLegend && hasSizeLegend && shouldShowPagingLegend(height, position)) {
+                        return (
+                            <ColorAndSizeLegendWithPaging
+                                {...props}
+                                numericSymbols={numericSymbols}
+                                availableLegends={availableLegends}
+                            />
+                        );
+                    }
+
+                    return renderColorAndSizeLegend(props, numericSymbols, availableLegends);
+                }}
+            </IntlTranslationsProvider>
+        </IntlWrapper>
+    );
+}
+
+function renderCategoryAndSizeLegend(
+    props: IGeoChartLegendRendererProps,
+    numericSymbols: string[],
+    availableLegends: IAvailableLegends,
+): JSX.Element {
+    const { hasSizeLegend } = availableLegends;
+    const classNames = getClassnames(props, availableLegends);
 
     return (
         <Measure client={true}>
             {({ measureRef, contentRect }: MeasuredComponentProps) => {
                 return (
                     <div className={classNames} ref={measureRef}>
-                        {renderPushpinColorLegend(props, hasColorLegend)}
-                        {renderPushpinCategoryLegend(props, contentRect, hasSizeLegend, hasCategoryLegend)}
-                        {renderPushpinSizeLegend(props, hasSizeLegend)}
+                        {renderPushpinCategoryLegend(props, contentRect, hasSizeLegend)}
+                        {renderPushpinSizeLegend(props, numericSymbols, hasSizeLegend)}
                     </div>
                 );
             }}
         </Measure>
     );
 }
 
-function renderPushpinColorLegend(props: IGeoChartLegendRendererProps, hasColorLegend: boolean): JSX.Element {
+// if height of color + size is bigger than container, we will show paging for legends
+function shouldShowPagingLegend(height: number, legendPosition: PositionType): boolean {
+    if (height !== undefined && (legendPosition === LEFT || legendPosition === RIGHT)) {
+        const heightOfColorAndSizeLegend = HEIGHT_OF_COLOR_LEGEND + HEIGHT_OF_SIZE_LEGEND;
+        return height < heightOfColorAndSizeLegend;
+    }
+    return false;
+}
+
+interface IColorAndSizeLegendWithPagingProps extends IGeoChartLegendRendererProps {
+    availableLegends: IAvailableLegends;
+    numericSymbols: string[];
+}
+
+interface IColorAndSizeLegendWithPagingState {
+    page: number;
+}
+
+class ColorAndSizeLegendWithPaging extends React.PureComponent<
```

Comment:
It could be like this
```
function GeoChartFluidLegend(props: IColorAndSizeLegendWithPagingProps): React.ReactElement {
    const [page, setPage] = useState<number>(1);
    
    const showNextPage = (): void => setPage(2);
    const showPrevPage = (): void => setPage(1);

    const { availableLegends, numericSymbols } = props;
    const classNames = getClassnames(props, availableLegends);
    return (
        <div className={classNames}>
            <div className="geo-legend-paging">
                {renderPushpinColorLegend(props, numericSymbols, page === 1)}
                {renderPushpinSizeLegend(props, numericSymbols, page === 2)}
            </div>
            <Paging
                page={page}
                pagesCount={2}
                showNextPage={showNextPage}
                showPrevPage={showPrevPage}
            />
        </div>
    );
}
```